### ISSUE TITLE
Promise handling algorithms are not very clear about what happens if type conversion fails

### ISSUE THREAD

**Issue by domenic** [heycam/webidl#782]
https://heycam.github.io/webidl/#dfn-perform-steps-once-promise-is-settled

What if the type is `Promise<Node>` and someone passes `Promise.resolve(3)`?

I think the best we can do is go down the rejection path? Although that makes using "upon fulfillment" in isolation pretty dangerous for non-`any` _T_s.

**Comment by domenic** [comment]
I am curious if implementations actually pay attention to the _T_ at all... Do they have counterpart algorithms?

**Comment by bzbarsky** [comment]
Gecko pays attention to the _T_ in the following ways:

1) IDL parse fails if `Promise<T>` is the return type of an operation or attribute but _T_ is not exposed in all the globals where the operation or attribute is exposed.  There is no corresponding requirement in the spec, as far as I can tell (also for the situation where some type is returned directly, without Promise being involved), but fundamentally it doesn't make sense to have this setup.
2) Something involving which C++ headers to include in generated binding files.  I'm not even convinced that code needs to be there.

There are no other places that are aware of the _T_ in Gecko as far as I can tell.

**Comment by annevk** [comment]
How does [`respondWith()`](https://w3c.github.io/ServiceWorker/#fetchevent-interface) work in Gecko? Is there a manual type check?

**Comment by Ms2ger** [comment]
In the spec as written, it'll return a rejected promise and the success steps won't be run.

Let's take a look at what APIs take a Promise argument anyway:

- [`ExtendableEvent.waitUntil`](https://w3c.github.io/ServiceWorker/#dom-extendableevent-waituntil): takes `Promise<any>`
- [`FetchEvent.respondWith`](https://w3c.github.io/ServiceWorker/#fetch-event-respondwith): takes `Promise<Response>`; if the conversion fails, it wants to do the same as with a rejected promise.
- [`WebAssembly.compileStreaming`](https://webassembly.github.io/spec/web-api/#dom-webassembly-compilestreaming) and [`WebAssembly.instantiateStreaming`](https://webassembly.github.io/spec/web-api/#dom-webassembly-instantiatestreaming) take `Promise<Response>` and want the currently specced behavior (implemented in Fx, Chrome & Edge per [wpt.fyi](https://wpt.fyi/results/wasm/webapi/invalid-args.any.html?label=master&label=experimental&aligned)). Rejection steps just pass the rejection value through.
- [`MerchantValidationEvent.complete`](https://w3c.github.io/payment-request/#complete-method-0): takes `Promise<any>`
- [`PaymentRequestUpdateEvent.updateWith`](https://w3c.github.io/payment-request/#updatewith-method) takes `Promise<PaymentDetailsUpdate>`; it wants to do nearly the same thing as when it gets a rejected promise, except that it drops the rejection value in a rejected promise in favour of an AbortError, while it passes along the TypeError from the conversion failure.

This suggests to me that we should run the rejection steps on conversion failure.

**Comment by bzbarsky** [comment]
> How does `respondWith()` work in Gecko? Is there a manual type check?

Yes.  See https://searchfox.org/mozilla-central/rev/7088fc958db5935eba24b413b1f16d6ab7bd13ea/dom/serviceworkers/ServiceWorkerEvents.cpp#576,588-591,595-597,605-608 (which checks that the resolution value is an object, then checks that it's a Response.

**Comment by domenic** [comment]
> In the spec as written, it'll return a rejected promise and the success steps won't be run.

Ah, I guess because "converting" will throw, and PerformPromiseThen handles it. That makes sense.

> This suggests to me that we should run the rejection steps on conversion failure.

Sounds good to me.
